dev/doc/parsing.md

```diff
@@ -53,111 +66,251 @@ which has the full grammar for Rocq
 
-* `ARGUMENT EXTEND` - defines a new nonterminal
+These constructs provide essential semantic information that's provided in a more complex,
+less readable way with `GRAMMAR EXTEND`.
+
+### VERNAC EXTEND
+
+`VERNAC COMMAND EXTEND` - adds new command syntax by adding productions to the
+`command` entry.
+
+`VERNAC { entry } EXTEND` - adds new command syntax to the `entry` entry.
+This is typically used for proof modes (see also `Pvernac.register_proof_mode` and the user doc for proof modes).
+
+Simple example:
+
+```
+VERNAC COMMAND EXTEND CmdName CLASSIFIED AS SIDEFF
+| [ "Cmd" arg(a) ] -> { do_interp a }
+END
+```
+defines a command composed of the string `Cmd` followed by argument `wit_arg`.
+Running the command runs `do_interp a` where `a` is the result of parsing `arg`.
 
-  ```
-  ARGUMENT EXTEND ast_closure_term
-       TYPED AS type_info
-       PRINTED BY { pp_ast_closure_term }
-       INTERPRETED BY { interp_ast_closure_term }
-       GLOBALIZED BY { glob_ast_closure_term }
-       SUBSTITUTED BY { subst_ast_closure_term }
-       RAW_PRINTED BY { pp_ast_closure_term }
-       GLOB_PRINTED BY { pp_ast_closure_term }
-    | [ term_annotation(a) constr(c) ] -> { mk_ast_closure_term a c }
-  END
-  ```
+Example with all optional elements:
+```
+VERNAC COMMAND EXTEND CmdName CLASSIFIED AS SIDEFF
+| #[ attr ] ![ state ] [ "Cmd" arg(a) ]
+  => { rule_classif a }
+  SYNTERP AS synv { do_synterp attr a }
+  -> { do_interp attr a synv }
+END
+```
+
+Productions here are represented with [alternate syntax](#alternate-production-syntax), described later.
 
-  See comments in `tacentries.mli` for partial information on the various
-  arguments.
+Nonterminals in `VERNAC EXTEND` parsing rules (e.g. `arg` in the example)
+are arguments, possibly with [nonterminal modifiers](#nonterminal-modifiers).
+The name prefixed by `wit_` must be available in the current OCaml scope.
 
-* `VERNAC ARGUMENT EXTEND` - (part of `argument_extend` in the yacc grammar) defines
-  productions for a single nonterminal.  For example:
+New commands should be added using this construct rather than `GRAMMAR EXTEND` so
+they are correctly registered, such as having the correct command classifier.
 
-  ```
-  VERNAC ARGUMENT EXTEND language
-  PRINTED BY { pr_language }
-  | [ "Ocaml" ] -> { let _ = warn_deprecated_ocaml_spelling () in Ocaml }
-  | [ "OCaml" ] -> { Ocaml }
-  | [ "Haskell" ] -> { Haskell }
-  | [ "Scheme" ] -> { Scheme }
-  | [ "JSON" ] -> { JSON }
-  END
-  ```
+#### Extend name
 
-  TODO: explain PRINTED BY, CODE
+Each `VERNAC EXTEND` must have a unique name ("CmdName" in the example)
+(unique for the plugin, you will get an anomaly at runtime if it's
+not).
 
-* DOC_GRAMMAR - Used in `doc_grammar`-generated files to permit simplified syntax
+#### Classification
 
-Note that you can reverse engineer many details by comparing the `.mlg` input file with
-the `.ml` generated by `coqpp`.
+Commands must be "classified" for the STM. This means generating a
```

Comment:
STM?

Reply:
"State Transaction Machine", it drives command execution in rocq compile, rocq top and coqidetop (not vsrocq and coq-lsp).
It is problematic so we would like to remove it someday but have not managed so far.